pypose/module/lqr.py

```diff
@@ -158,41 +249,53 @@ def __init__(self, system, Q, p, T):
         assert self.Q.device == self.p.device, "device not compatible."
         assert self.Q.dtype == self.p.dtype, "tensor data type not compatible."
 
-    def forward(self, x_init):
+    def forward(self, x_init, current_x, current_u, time):
         r'''
         Performs LQR for the linear system.
 
         Args:
             x_init (:obj:`Tensor`): The initial state of the system.
+            current_x (:obj:`Tensor`, optinal): The current states of the system along a
+                trajectory.
+            current_u (:obj:`Tensor`, optinal): The current inputs of the system along a
+                trajectory.
+            time (:obj:`Tensor`, optinal): The reference time step of the dynamical
+                system.
 
         Returns:
             List of :obj:`Tensor`: A list of tensors including the solved state sequence
             :math:`\mathbf{x}`, the solved input sequence :math:`\mathbf{u}`, and the associated
             quadratic costs :math:`\mathbf{c}` over the time horizon.
         '''
-        K, k = self.lqr_backward()
-        x, u, cost = self.lqr_forward(x_init, K, k)
+        K, k = self.lqr_backward(current_x, current_u, time)
+        x, u, cost = self.lqr_forward(x_init, current_x, current_u, K, k)
         return x, u, cost
 
-    def lqr_backward(self):
+    def lqr_backward(self, current_x, current_u, time):
+
+        # Q: (n_batch*, T, N, N), p: (n_batch*, T, N), where n_batch* can be any batch dimensions, e.g., (2, 3)
+        n_batch = self.p.shape[:-2]
+        ns, nc = current_x.size(-1), current_u.size(-1)
```

Comment:
I noticed these appeared multiple time. Why don't you just obtain these once at the beginning of the __init_ to make them the class member. Because one the system is passed through the class, ns, nc should be fixed. There is no need to compute them several times.